src/transforms/route.rs

```diff
@@ -16,48 +16,127 @@ use crate::{
+    for (output_name, condition) in config.route.iter() {
+        let condition = condition.build(&context.enrichment_tables)?;
+        conditions.push((output_name.clone(), condition));
+    }
+    Ok(conditions)
+}
+
+//------------------------------------------------------------------------------
+
 #[derive(Clone)]
-pub struct Route {
+pub struct EveryMatchRoute {
     conditions: Vec<(String, Condition)>,
 }
 
-impl Route {
+impl EveryMatchRoute {
     pub fn new(config: &RouteConfig, context: &TransformContext) -> crate::Result<Self> {
-        let mut conditions = Vec::with_capacity(config.route.len());
-        for (output_name, condition) in config.route.iter() {
-            let condition = condition.build(&context.enrichment_tables)?;
-            conditions.push((output_name.clone(), condition));
-        }
-        Ok(Self { conditions })
+        Ok(Self {
+            conditions: build_conditions(config, context)?,
+        })
     }
 }
 
-impl SyncTransform for Route {
+impl SyncTransform for EveryMatchRoute {
     fn transform(
         &mut self,
         event: Event,
         output: &mut vector_core::transform::TransformOutputsBuf,
     ) {
+        let mut discarded = Vec::with_capacity(self.conditions.len());
         for (output_name, condition) in &self.conditions {
             if condition.check(&event) {
                 output.push_named(output_name, event.clone());
             } else {
-                emit!(&RouteEventDiscarded {
-                    output: output_name.as_ref()
-                })
+                discarded.push(output_name.as_ref());
             }
         }
+        if discarded.len() == self.conditions.len() {
+            output.push(event);
+        }
+        if !discarded.is_empty() {
+            emit!(&RouteEventDiscarded { outputs: discarded });
+        }
     }
 }
 
 //------------------------------------------------------------------------------
 
+#[derive(Clone)]
+pub struct FirstMatchRoute {
+    conditions: Vec<(String, Condition)>,
+}
+
+impl FirstMatchRoute {
+    pub fn new(config: &RouteConfig, context: &TransformContext) -> crate::Result<Self> {
+        Ok(Self {
+            conditions: build_conditions(config, context)?,
+        })
+    }
+}
+
+impl SyncTransform for FirstMatchRoute {
+    fn transform(
+        &mut self,
+        event: Event,
+        output: &mut vector_core::transform::TransformOutputsBuf,
+    ) {
+        if let Some((output_name, _)) = self
+            .conditions
+            .iter()
+            .filter(|(_, condition)| condition.check(&event))
+            .next()
+        {
+            output.push_named(output_name, event.clone());
+        } else {
+            emit!(&RouteEventDiscarded {
+                outputs: self
+                    .conditions
+                    .iter()
+                    .map(|(name, _)| name.as_str())
+                    .collect()
+            });
```

Comment:
If we expect this to be emitted frequently, could this resulting vec be cached? It will be identical for any given config.

Reply:
Same as above